### PR TITLE
Converter to IOB/BIO + code/docs cleanup

### PR DESCRIPTION
**Pull request by gawy**
* Moved in logic for data conversion from Stanza and stanza-lang-uk repositories. IOB/BIO files can now be created here
* Removed dependency on https://github.com/gawy/stanza-lang-uk
* fix: Added progress bar for MITIE artifact download process

### CODE REVIEW

File: scripts/convert_data.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    :return: data in BEIOS or IOB format https://en.wikipedia.org/wiki/Inside–outside–beginning_(tagging)
+    """
+
+    def join_simple_chunk(chunk: str) -> list:
+        if len(chunk.strip()) == 0:
+            return []
+        # keep the newlines, but discard the non-newline whitespace
+        tokens = re.split(r'(\n)|\s', chunk.strip())
+        # the re will return None for splits which were not caught in a group
+        tokens = [x for x in tokens if x is not None]
```
Comment (dchaplinsky) [inline scripts/convert_data.py:69]: You might use `filter(None, tokens)` (don't forget that it returns iterator, not the list)

File: scripts/convert_data.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    converters = {'beios': format_token_as_beios, 'iob': format_token_as_iob}
+    res = []
+    markup = parse_bsf(bsf_markup)
+
+    prev_idx = 0
+    m_ln: BsfInfo
+    for m_ln in markup:
+        res += join_simple_chunk(data[prev_idx:m_ln.start_idx])
+
+        convert_f = converters[converter]
```
Comment (dchaplinsky) [inline scripts/convert_data.py:81]: You might move that outside of the loop, this will save some lookups (as long as converter variable doesn't change)

File: scripts/convert_data.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        res.extend(convert_f(m_ln.token, m_ln.tag))
+        prev_idx = m_ln.end_idx
+
+    if prev_idx < len(data) - 1:
+        res += join_simple_chunk(data[prev_idx:])
+
+    return '\n'.join(res)
+
+
+def parse_bsf(bsf_data: str) -> list:
```
Comment (dchaplinsky) [inline scripts/convert_data.py:91]: Any good ideas how to keep this function and `BsfInfo` definition in one place? Separate package looks like an overkill, of course.

Comment (dchaplinsky) [reply scripts/convert_data.py:91]: What makes me a bit nervous is that complex regex, that we now have in two different packages. On the other hand I don't think it's going to change anytime soon :)

Comment (dchaplinsky) [reply scripts/convert_data.py:91]: Maybe (just maybe), the whole logic to convert between formats can be separated as a package, which has some value per se (i.e might be used by other people around the world).

File: scripts/convert_data.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    :param src_dir_path: path to directory with BSF marked files
+    :param dst_dir_path: where to save output data
+    :param converter: `beios` or `iob` output formats
+    :param train_test_split_file: path to file containing train/test lists of file names
+    :return:
+    """
+    # following 2 constants need to comply with stanza naming for corpus and language
+    corpus_name = 'Ukrainian-languk'
+
+    ann_path = os.path.join(src_dir_path, '*.tok.ann')
```
Comment (dchaplinsky) [inline scripts/convert_data.py:123]: Again, as a general recommendation (as long as we are using Python3), `pathlib` is a very nice abstraction on top of os.path.join, and you can use `type=pathlib.Path` in argparse) Just don't forget to convert to str, when feeding it to `glob.glob`

File: scripts/convert_data.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+    tok_path = os.path.join(src_dir_path, '*.tok.txt')
+    tok_files = glob.glob(tok_path)
+    tok_files.sort()
+
+    corpus_folder = os.path.join(dst_dir_path, corpus_name)
+    if not os.path.exists(corpus_folder):
+        os.makedirs(corpus_folder)
+
+    if len(ann_files) == 0 or len(tok_files) == 0:
```
Comment (dchaplinsky) [inline scripts/convert_data.py:135]: you can just do `if not ann_files or not tok_files:`

File: scripts/convert_data.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        log.warning(f'Token and annotation files are not found at specified path {ann_path}')
+        return
+    if len(ann_files) != len(tok_files):
```
Comment (dchaplinsky) [inline scripts/convert_data.py:138]: What is the consequence of that? Should we just stop?

File: scripts/convert_data.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            tqdm.write(f'Token and Annotation file names do not match ann={ann_fname}, tok={tok_fname}')
+            continue
+
+        with open(tok_fname) as tok_file, open(ann_fname) as ann_file:
+            token_data = tok_file.read()
+            ann_data = ann_file.read()
+            out_data = convert_bsf(token_data, ann_data, converter)
+
+            if train_test_split_file is None:
+                target_dataset = choices(data_sets, split_weights)[0]
```
Comment (dchaplinsky) [inline scripts/convert_data.py:164]: We probably should add an option to provide the random seed as a param somewhere, so we can have reliable splits

File: scripts/convert_data.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            else:
+                target_dataset = train_set
+                fkey = os.path.basename(tok_fname)[:-4]
```
Comment (dchaplinsky) [inline scripts/convert_data.py:167]: `os.path.splitext` seems to be a more robust alternative

File: scripts/tokenize_uk.py [hunk 9]
```diff
@@ -17,7 +17,7 @@ def tokenize_text(string):
     for part in string.split(u'\n'):
         par = []
         for sent in tokenize_sents(part):
-            par.append(tokenize_words(string))
+            par.append(tokenize_words(sent))
```
Comment (dchaplinsky) [inline scripts/tokenize_uk.py:20]: Ouch! Eagle eye!

File: scripts/train_mitie_ner.py [hunk 10]
```diff
@@ -66,9 +71,20 @@ def prepare_mitie_training_data(dev_files):
 
 
 def download_file(download_url, file_name):
-    with requests.get(download_url, stream=True) as r:
-        with open(file_name, 'wb') as f:
-            shutil.copyfileobj(r.raw, f)
+    resp = requests.get(download_url, stream=True)
+    total = int(resp.headers.get('content-length', 0))
+
+    with open(file_name, 'wb') as file, tqdm(
```
Comment (dchaplinsky) [inline scripts/train_mitie_ner.py:77]: Sweeet!

File: scripts/train_mitie_ner.py [hunk 11]
```diff
@@ -90,8 +106,7 @@ def run_training(cpu_threads, config_path, feature_extractor_path):
 
         if not os.path.exists(feature_extractor_path):
             print(f'Feature extractor file not provided or not found. '
-                  f'\nTrying to download from {feature_extractor_url}'
-                  f'\nSorry no progress bar.')
```
Comment (dchaplinsky) [inline scripts/train_mitie_ner.py:94]: Even sweeter!

**Review by dchaplinsky — APPROVED** [review]
Once again, thanks for the all the hard work done (and tests)!

My comments aren't mandatory, feel free to merge as is.
One potential point for discussion is the idea to separate all the conversion routines into a separate package so everyone can use it. But again, I'm insisting on that at all.

File: scripts/convert_data.py (hunk 3 above)
Comment (gawy) [reply scripts/convert_data.py:91]: I also spent quite a bit of time debating myself whether I should package this as a library. Not a fan of duplicating code but ended up with the following:
- so far this code is generally unique to our 2 repos
- adding an external dependency because 1 function might be an overkill and complicate dependency management and making changes in that shared code
- makes a bit harder to read and understand this relatively simple code
Happy to package it as soon as we have a need for 3rd usage :)

File: scripts/convert_data.py (hunk 1 above)
Comment (gawy) [reply scripts/convert_data.py:69]: Dove deep into the topic - looks like a religious debate :)
https://stackoverflow.com/questions/3013449/list-comprehension-vs-lambda-filter
I Will keep it as is - as this is also the code moved from Stanza

File: scripts/convert_data.py (hunk 2 above)
Comment (gawy) [reply scripts/convert_data.py:81]: Good point thanks

File: scripts/convert_data.py (hunk 6 above)
Comment (gawy) [reply scripts/convert_data.py:138]: That's a health check. Added it just in case there are some misalignments in data set. So we can at least see that something went wrong. It won't break the logic but the zip() function in further for loop will just iterate on the shortest list.

After looking at it once more - if such a situation happens - the data set will probably be a mess as there is a high chance that .tok and .ann will be misaligned.
In such case we will at least see a log.
Changing it to error level and actually adding return - no point in just printing the message.

File: scripts/convert_data.py (hunk 8 above)
Comment (gawy) [reply scripts/convert_data.py:168]: Makes sense. Fixed

File: scripts/convert_data.py (hunk 7 above)
Comment (gawy) [reply scripts/convert_data.py:165]: We already have split file and in case someone wants a random split - let it be random. What do you think?

Comment (dchaplinsky) [reply scripts/convert_data.py:166]: Sometime, it has a point to have some control over randomness. But as I said, it's not a mandatory point in any sense.

File: scripts/convert_data.py (hunk 3 above)
Comment (dchaplinsky) [reply scripts/convert_data.py:92]: Ok, works for me

**Comment by dchaplinsky** [comment]
Happy to merge. Two misc comments that doesn't really important, happy to ignore them.
